pool/src/lib.rs

```diff
@@ -690,47 +844,67 @@ impl LendingPool {
             let curr_reserve = read_reserve(env, curr_reserve_asset.clone())?;
 
+            if !curr_reserve.configuration.is_active && liquidation {
+                return Err(Error::NoActiveReserve);
+            }
+
             let reserve_price = Self::get_asset_price(env, curr_reserve_asset.clone())?;
 
             if user_config.is_using_as_collateral(env, i) {
                 let coll_coeff = Self::get_collateral_coeff(env, &curr_reserve)?;
 
                 let who_balance: i128 = match mb_who_balance.clone() {
-                    Some((address, balance)) if address == curr_reserve.s_token_address.clone() => {
+                    Some(AssetBalance { asset, balance })
+                        if asset == curr_reserve.s_token_address.clone() =>
+                    {
                         balance
                     }
                     _ => STokenClient::new(env, &curr_reserve.s_token_address).balance(&who),
                 };
 
                 let discount = FixedI128::from_percentage(curr_reserve.configuration.discount)
                     .ok_or(Error::CalcAccountDataMathError)?;
-                let compounded_balance = discount
-                    .mul_int(
-                        coll_coeff
-                            .mul_int(who_balance)
-                            .ok_or(Error::CalcAccountDataMathError)?,
-                    )
+
+                let compounded_balance = coll_coeff
+                    .mul_int(who_balance)
                     .ok_or(Error::CalcAccountDataMathError)?;
 
-                let liquidity_balance_in_xlm = reserve_price
+                let compounded_balance_in_xlm = reserve_price
                     .mul_int(compounded_balance)
                     .ok_or(Error::CalcAccountDataMathError)?;
 
-                total_collateral_in_xlm = total_collateral_in_xlm
-                    .checked_add(liquidity_balance_in_xlm)
+                let discounted_balance_in_xlm = discount
+                    .mul_int(compounded_balance_in_xlm)
                     .ok_or(Error::CalcAccountDataMathError)?;
-            }
 
-            if user_config.is_borrowing(env, i) {
+                total_discounted_collateral_in_xlm = total_discounted_collateral_in_xlm
+                    .checked_add(discounted_balance_in_xlm)
+                    .ok_or(Error::CalcAccountDataMathError)?;
+
+                if liquidation {
+                    let curr_discount = curr_reserve.configuration.discount;
+                    let mut collateral_to_receive = sorted_collateral_to_receive
+                        .get(curr_discount)
+                        .unwrap_or(Ok(Vec::new(env)))
+                        .expect("sorted");
+                    collateral_to_receive.push_back((
+                        curr_reserve,
+                        who_balance,
+                        reserve_price.into_inner(),
+                        coll_coeff.into_inner(),
```

Comment:
It's better to use FixedI128 type for price and coeff here, to avoid conversation into/from_inner